File: Search/Medium/search_an_element_in_sorted_and_rotated_array.py
#User function Template for python3
# https://practice.geeksforgeeks.org/problems/search-in-a-rotated-array0959/1
# Given a sorted and rotated array A of N distinct elements which is rotated at some point, and given an 
# element K. The task is to find the index of the given element K in the array A.
def Search(arr,n,k):
    l, r= 0, n-1
    
    while(l <= r):
        mid = int(l + (r - l)//2)
        if arr[mid] == k:
            return mid
            
        elif arr[mid] < k:
            if k > arr[r] and arr[r] > arr[mid]:
                r = mid -1
            else :
                l = mid + 1
                
        else:
            if k < arr[l] and arr[l] < arr[mid]:
                l = mid + 1
            else:
                r = mid - 1
                
    return -1


#{ 
 # Driver Code Starts
#Initial Template for Python 3

if __name__ == '__main__':
    tcs=int(input())
    for _ in range(tcs):
        n=int(input())
        arr=[int(x) for x in input().split()]
        k=int(input())

        print(Search(arr,n,k))

# } Driver Code Ends

File: Search/Medium/search_an_element_in_sorted_and_rotated_array.cpp
//{ Driver Code Starts
//Initial template for C++

#include<bits/stdc++.h>
using namespace std;

int Search(vector<int> ,int );

//User code will be pasted here

int main(){

    int t;
    cin >> t;

    while(t--){
        int n;
        cin >> n;
        
        vector<int> vec(n);
        
        for(int i =0;i<n;i++) 
            cin >> vec[i];
        
        int target;
        cin >> target;
        
        cout << Search(vec,target) << "\n";
         
    }
}
// } Driver Code Ends

// https://practice.geeksforgeeks.org/problems/search-in-a-rotated-array0959/1
// Given a sorted and rotated array A of N distinct elements which is rotated at some point, and given an 
// element K. The task is to find the index of the given element K in the array A.
//User function template for C++

// vec : given vector of elements
// K : given value whose index we need to find 
int Search(vector<int> vec, int k) {
    int l = 0, r = vec.size()-1;
    while(l <= r)
    {
        int mid = (l + r)/ 2;
        if(vec[mid] == k) return mid;
        else if(vec[mid] < k)
        {
            if(k > vec[r] && vec[r] > vec[mid]) r = mid - 1;
            else l = mid + 1;
        }
        else 
        {
            if(k < vec[l] && vec[l] < vec[mid]) l = mid + 1;
            else r = mid - 1;
        }
    }
    
    return -1;
}